runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java

```diff
@@ -90,4 +113,161 @@ public void healNode(@Nonnull String endpointToHeal, int retry, @Nonnull Duratio
+                            = new ArrayList<>(layoutStripe.getLogServers());
+                    // Retain only the responsive servers.
+                    responsiveLogServers.retainAll(peerResponsiveNodes);
+
+                    if (!responsiveLogServers.containsAll(layoutStripe.getLogServers())) {
+                        if (clusterStatus.equals(ClusterStatus.STABLE)) {
+                            clusterStatus = ClusterStatus.DEGRADED;
+                        }
+                        int quorumSize = (layoutStripe.getLogServers().size() / 2) + 1;
+                        if (responsiveLogServers.size() < quorumSize) {
+                            clusterStatus = ClusterStatus.UNAVAILABLE;
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        return clusterStatus;
+    }
+
+    /**
+     * Gets the cluster status by pinging each node.
+     * These pings are then compared to the layout to decide whether the cluster is:
+     * STABLE, DEGRADED OR UNAVAILABLE.
+     * The report consists of the following:
+     * Layout - at which the report was generated.
+     * ClientServerConnectivityStatusMap - the connectivity status of this client to the cluster.
+     * ClusterStatus - Health of the Corfu cluster.
+     *
+     * @return ClusterStatusReport
+     */
+    public ClusterStatusReport getClusterStatus() {
+        RuntimeLayout l = new RuntimeLayout(runtime.getLayoutView().getLayout(), runtime);
+        Layout layout = l.getLayout();
+        // All configured nodes.
+        Set<String> allNodes = new HashSet<>(l.getLayout().getAllServers());
+
+        // Counters to track heartbeat responses from nodes.
+        // A counter is incremented even if we get a WrongEpochException as the node is alive.
+        Map<String, Integer> counters = new HashMap<>();
+        allNodes.forEach(endpoint -> counters.put(endpoint, 0));
+
+        // Map of nodeView received from heartbeatResponses.
+        Map<String, NodeView> nodeViewMap = new HashMap<>();
+
+        // Ping all nodes CLUSTER_STATUS_QUERY_ATTEMPTS times.
+        for (int i = 0; i < CLUSTER_STATUS_QUERY_ATTEMPTS; i++) {
+            Map<String, CompletableFuture<NodeView>> futureMap = new HashMap<>();
+
+            // Ping all nodes asynchronously
+            for (String endpoint : allNodes) {
+                CompletableFuture<NodeView> cf = new CompletableFuture<>();
+                try {
+                    cf = l.getManagementClient(endpoint).sendHeartbeatRequest();
+                } catch (Exception e) {
+                    // Requesting the heartbeat can cause NetworkException if connection cannot
+                    // be established.
+                    cf.completeExceptionally(e);
+                }
+                futureMap.put(endpoint, cf);
+            }
+
+            // Accumulate all responses.
+            for (String endpoint : futureMap.keySet()) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Iterate over the "entrySet" instead of the "keySet". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2864)